pkg/pool-weighted/test/WeightedPool.test.ts

```diff
@@ -170,4 +170,108 @@ describe('WeightedPool', function () {
+    let vaultContract: Contract;
+
+    sharedBeforeEach('deploy pool', async () => {
+      tokens = allTokens.subset(numTokens);
+      const vault = await Vault.create();
+      vaultContract = vault.instance;
+
+      pool = await WeightedPool.create({
+        poolType: WeightedPoolType.WEIGHTED_POOL,
+        tokens,
+        weights: WEIGHTS.slice(0, numTokens),
+        swapFeePercentage: swapFeePercentage,
+        vault,
+      });
+      await vault.setSwapFeePercentage(protocolFeePercentage);
+
+      await pool.updateProtocolFeePercentageCache();
+    });
+
+    context('before initialized', () => {
+      it('rate is zero', async () => {
+        await expect(pool.getRate()).to.be.revertedWith('ZERO_INVARIANT');
+      });
+    });
+
+    context('once initialized', () => {
+      sharedBeforeEach('initialize pool', async () => {
+        // Init pool with equal balances so that each BPT accounts for approximately one underlying token.
+        const equalBalances = Array(numTokens).fill(fp(100));
+
+        await allTokens.mint({ to: lp.address, amount: fp(1000) });
+        await allTokens.approve({ from: lp, to: pool.vault.address });
+
+        await pool.init({ from: lp, recipient: lp.address, initialBalances: equalBalances });
+      });
+
+      context('without protocol fees', () => {
+        it('reports correctly', async () => {
+          const totalSupply = await pool.totalSupply();
+          const invariant = await pool.estimateInvariant();
+
+          const expectedRate = fpDiv(invariant.mul(numTokens), totalSupply);
+          const rate = await pool.getRate();
+
+          expect(rate).to.be.equalWithError(expectedRate, 0.0001);
+        });
+      });
+
+      context('with protocol fees', () => {
+        let originalRate: BigNumber;
+
+        sharedBeforeEach('swap bpt in', async () => {
+          const amount = fp(20);
+          const tokenIn = tokens.first;
+          const tokenOut = tokens.second;
+
+          originalRate = await pool.getRate();
+
+          const singleSwap = {
+            poolId: await pool.getPoolId(),
+            kind: SwapKind.GivenIn,
+            assetIn: tokenIn.address,
+            assetOut: tokenOut.address,
+            amount: amount,
+            userData: '0x',
+          };
+
+          const funds: FundManagement = {
+            sender: lp.address,
+            recipient: lp.address,
+            fromInternalBalance: false,
+            toInternalBalance: false,
+          };
+
+          await vaultContract.connect(lp).swap(singleSwap, funds, 0, MAX_UINT256);
+        });
+
+        it("doesn't include the value of uncollected protocol fees in the rate", async () => {
+          const rate = await pool.getRate();
+
+          // The rate considering fees should be lower. Check that we have a difference of at least 0.01% to discard
+          // rounding error.
+          expect(originalRate).to.be.lt(rate.mul(9999).div(10000));
+        });
```

Comment:
It isn't clear how the description follows from the test here. We're checking that swap fees increase the rate but not that it only counts the non-protocol swap fees.

Reply:
I think that for this test we should be calculating the rate based on the balances and ignoring any protocol fees and show that the rate is significantly below this value.